docs/site/Context.md

```diff
@@ -229,3 +229,150 @@ class HelloController {
+  process
+  - err: the error object thrown
+
+When an existing binding key is replaced with a new one, an `unbind` event is
+emitted for the existing binding followed by a `bind` event for the new binding.
+
+## Context observers
+
+Bindings can be added or removed to a context object. With emitted context
+events, we can add listeners to a context object to be invoked when bindings
+come and go. There are a few caveats associated with that:
+
+1. The binding object might not be fully configured when a `bind` event is
+   emitted.
+
+   For example:
+
+   ```ts
+   const ctx = new Context();
+   ctx
+     .bind('foo')
+     .to('foo-value')
+     .tag('foo-tag');
+   ctx.on('bind', binding => {
+     console.log(binding.tagNames); // returns an empty array `[]`
+   });
+   ```
+
+   The context object emits a `bind` event when `ctx.bind` method is called. It
+   does not control the fluent apis `.to('foo-value').tag('foo-tag')`, which
+   happens on the newly created binding object. As a result, the `bind` event
+   listener receives a binding object which only has the binding key populated.
+
+   A workaround is to create the binding first before add it to a context:
+
+   ```ts
+   const ctx = new Context();
+   const binding = Binding.create('foo')
+     .to('foo-value')
+     .tag('foo-tag');
+   ctx.add(binding);
+   ctx.on('bind', binding => {
+     console.log(binding.tagMap); // returns `['foo-tag']`
+   });
+   ```
+
+2. It's hard for event listeners to perform asynchronous operations.
+
+To make it easy to support asynchronous event processing, we introduce
+`ContextObserver` and corresponding APIs on `Context:
+
+1. ContextObserver interface
+
+```ts
+/**
+ * Observers of context bind/unbind events
+ */
+export interface ContextObserver {
+  /**
+   * An optional filter function to match bindings. If not present, the listener
+   * will be notified of all binding events.
+   */
+  filter?: BindingFilter;
+
+  /**
+   * Listen on `bind`, `unbind`, or other events
+   * @param eventType Context event type
+   * @param binding The binding as event source
+   */
+  observe(
+    eventType: ContextEventType,
+    binding: Readonly<Binding<unknown>>,
+    context: Context,
+  ): ValueOrPromise<void>;
+}
+```
+
+2. Context APIs
+
+- `subscribe(observer: ContextObserver)`
+
+  Add a context event observer to the context chain, including its ancestors
+
+- `unsubscribe(observer: ContextObserver)`
+
+  Remove the context event observer from the context chain
+
+- `close()`
+
+  Close the context and release references to other objects in the context
+  chain. Please note observers of a child context are also registered with its
+  parent context. As a result, the `close` method must be called to avoid memory
+  leak if the child context is to be recycled.
+
+To react on context events asynchronously, we need to implement the
+`ContextObserver` interface and register it with the context.
```

Comment:
Please add a short description to explain the difference in how are ContextObserver instances invoked when compared to regular events:
- observers are called in the next turn of Promise micro-queue
- when there are multiple async observers registered, do they run in parallel or in series?
- when there are multiple events triggered in the same event loop and there are async observers registered, in which order will they be triggered?

Reply:
Added